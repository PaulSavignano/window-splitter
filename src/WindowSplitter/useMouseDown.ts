import {
  useCallback,
  useEffect,
  useRef,
} from 'react'

function useSplitter({ orientation = 'vertical' }) {
  const activePreviousElementRef = useRef<HTMLDivElement | null>(null)
  const activeResizerRef = useRef<HTMLDivElement | null>(null)
  const activeSetStyle = useRef(({ }) => { })
  const isResizing = useRef(false)

  const onMouseMove = useCallback((e) => {
    if (!isResizing.current || !activeResizerRef.current || !activePreviousElementRef.current) return
    switch (orientation) {
      case 'vertical': {
        const offsetTop = activePreviousElementRef.current.offsetTop
        const height = e.clientY - offsetTop
        return activeSetStyle.current({ height })
      }
      case 'horizontal': {
        const offsetLeft = activePreviousElementRef.current.offsetLeft
        const width = e.clientX - offsetLeft
        return activeSetStyle.current({ minWidth: width })
      }
      default:
        throw Error('Orientation not provided')
    }
  }, [orientation])

  const onMouseDown = useCallback(({ resizerRef, previousElementRef, setPreviousElementStyle }) => {
    activePreviousElementRef.current = previousElementRef.current
    activeResizerRef.current = resizerRef.current
    activeSetStyle.current = setPreviousElementStyle
    isResizing.current = true
  }, [])

  const onMouseUp = useCallback((e) => {
    activePreviousElementRef.current = null
    activeResizerRef.current = null
    activeSetStyle.current = ({ }) => { }
    isResizing.current = false
  }, [])

  useEffect(() => {
    document.addEventListener('mouseup', onMouseUp);
    document.addEventListener('mousemove', onMouseMove);
    return () => {
      document.removeEventListener('mouseup', onMouseUp);
      document.removeEventListener('mousemove', onMouseMove);
    }
  }, [onMouseMove, onMouseUp])

  return onMouseDown
}

export default useSplitter