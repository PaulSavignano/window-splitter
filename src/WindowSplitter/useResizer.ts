import {
  useMemo,
  useRef,
  useState,
} from 'react'

interface State {
  flex?: string,
  height?: number,
  minWidth?: number
  width?: number,
}

const initialState: State = {
  flex: '1 1 auto'
}

interface UseResizer {
  onMouseDown: Function
}

const useResizer = ({ onMouseDown }: UseResizer) => {
  const [previousElementStyle, setPreviousElementStyle] = useState(initialState)
  const resizerRef = useRef<HTMLDivElement>(null)
  const previousElementRef = useRef<HTMLDivElement>(null)
  return useMemo(() => ({
    previousElementStyle,
    previousElementRef,
    resizerElementProps: {
      ref: resizerRef,
      onMouseDown: () => onMouseDown({ resizerRef, previousElementRef, setPreviousElementStyle }),
    },
  }), [onMouseDown, previousElementStyle])
}

export default useResizer