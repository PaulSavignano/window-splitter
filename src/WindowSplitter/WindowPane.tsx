import React from 'react'

import useResizer from './useResizer'

interface WindowProps {
  child: React.ReactNode
  onMouseDown: Function
  dividerClassName: string
}

const Window: React.FC<WindowProps> = ({ child, onMouseDown, dividerClassName }) => {
  const { resizerElementProps, previousElementStyle, previousElementRef } = useResizer({ onMouseDown })
  const Child = React.cloneElement(child as React.ReactElement<any>, { ref: previousElementRef, style: previousElementStyle })
  return (
    <>
      {Child}
      <div {...resizerElementProps} role="separator" className={dividerClassName} />
    </>
  )
}

export default Window