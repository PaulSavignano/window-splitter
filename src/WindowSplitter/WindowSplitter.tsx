import React from 'react'
import { makeStyles, createStyles } from '@material-ui/core/styles'

import WindowPane from './WindowPane'
import useMouseDown from './useMouseDown'

const useStyles = makeStyles((theme) =>
  createStyles({
    horizontal: {
      display: 'flex',
      flexFlow: 'row',
      overflowY: 'auto',
    },
    vertical: {
      display: 'flex',
      flexFlow: 'column',
      height: '100%'
    },
    dividerHorizontal: {
      cursor: 'col-resize',
      border: '2px solid black',
    },
    dividerVertical: {
      cursor: 'row-resize',
      width: '100%',
      border: '2px solid black',
    }
  }),
);


interface WindowSplitterProps {
  orientation: string
  className?: string,
  style?: { [key: string]: string | number }
}

interface ObjectMap {
  [key: string]: string
}

const classNames = (...args: any[]) => args.filter(Boolean).join(' ')

const WindowSplitter: React.FC<WindowSplitterProps> = ({
  children,
  className,
  style,
  orientation,
}) => {
  const onMouseDown = useMouseDown({ orientation })
  const classes: ObjectMap = useStyles()
  const dividerClassName = orientation === 'horizontal' ? classes.dividerHorizontal : classes.dividerVertical
  return (
    <div
      className={classNames(classes[orientation], className)}
      style={style}
    >
      {React.Children.map(children, (child: React.ReactNode) => (
        <WindowPane
          child={child}
          onMouseDown={onMouseDown}
          dividerClassName={dividerClassName}
        />
      ))}
    </div>
  )
}

export default WindowSplitter
