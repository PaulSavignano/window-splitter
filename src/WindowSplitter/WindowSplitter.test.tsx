import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import { fireEvent, render } from '@testing-library/react'

import WindowSplitter from './WindowSplitter'

describe('WindowSplitter', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(
      <WindowSplitter orientation="horizontal">
        <article>One</article>
        <article>Two</article>
        <article>Three</article>
      </WindowSplitter>
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it('should set width', () => {
    const { getAllByRole, getByText } = render(
      <WindowSplitter orientation="horizontal">
        <article>One</article>
        <article>Two</article>
        <article>Three</article>
      </WindowSplitter>
    )
    const divider = getAllByRole('separator')[0]
    fireEvent.mouseDown(divider)
    fireEvent.mouseMove(divider, { clientX: 32 })
    const article = getByText('One')
    expect(article).toHaveStyle(`min-width: 32px`)
  })
})