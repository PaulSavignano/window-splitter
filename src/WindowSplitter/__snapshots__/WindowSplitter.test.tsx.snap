// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WindowSplitter should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-horizontal-1"
  >
    <article
      style="flex: 1 1 auto;"
    >
      One
    </article>
    <div
      class="makeStyles-dividerHorizontal-3"
      role="separator"
    />
    <article
      style="flex: 1 1 auto;"
    >
      Two
    </article>
    <div
      class="makeStyles-dividerHorizontal-3"
      role="separator"
    />
    <article
      style="flex: 1 1 auto;"
    >
      Three
    </article>
    <div
      class="makeStyles-dividerHorizontal-3"
      role="separator"
    />
  </div>
</DocumentFragment>
`;
